src/gwt/panmirror/src/editor/src/behaviors/insert_symbol/insert_symbol-popup.tsx

```diff
@@ -0,0 +1,188 @@
+/*
+ * insert_symbol-popup.tsx
+ *
+ * Copyright (C) 2019-20 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+import { WidgetProps } from '../../api/widgets/react';
+import React, { ChangeEvent } from 'react';
+import SymbolDataManager, { kCategoryAll, SymbolCharacter, SymbolCharacterGroup } from './insert_symbol-data';
+import { isElementFocused, focusElement } from '../../api/focus';
+import { Popup } from '../../api/widgets/popup';
+import { TextInput } from '../../api/widgets/text';
+import { SelectInput } from '../../api/widgets/select';
+import SymbolCharacterGrid, { newIndexForKeyboardEvent } from './insert_symbol-grid';
+
+import './insert_symbol-styles.css';
+
+const symbolDataManager = new SymbolDataManager();
+
+interface InsertSymbolPopupProps extends WidgetProps {
+  enabled: boolean;
+  size: [number, number];
+  searchImage?: string;
+  onInsertText: (text: string) => void;
+  onClose: VoidFunction;
+}
+
+export const InsertSymbolPopup: React.FC<InsertSymbolPopupProps> = props => {
+  const kPopupChromeHeight = 25;
+  const popupHeight = props.size[0] - kPopupChromeHeight;
+  const popupWidth = props.size[1];
+  const style: React.CSSProperties = {
+    ...props.style,
+    height: popupHeight + 'px',
+    width: popupWidth + 'px',
+  };
+
+  const gridHeight = popupHeight - 48;
+  const gridWidth = popupWidth;
+  const kNumberOfcolumns = 12;
+
+  const [filterText, setFilterText] = React.useState<string>('');
+  const [selectedSymbolGroup, setSelectedSymbolGroup] = React.useState<string>(kCategoryAll);
+  const [selectedSymbolIndex, setSelectedSymbolIndex] = React.useState<number>(0);
+  const [symbols, setSymbols] = React.useState<SymbolCharacter[]>([]);
+  const [filteredSymbols, setFilteredSymbols] = React.useState<SymbolCharacter[]>(symbols);
+
+  React.useEffect(() => {
+    const symbolsForGroup: SymbolCharacter[] = symbolDataManager.getSymbols(selectedSymbolGroup);
+    setSymbols(symbolsForGroup);
+  }, [selectedSymbolGroup]);
+
+  React.useEffect(() => {
+    setFilteredSymbols(symbolDataManager.filterSymbols(filterText, symbols));
+  }, [filterText, symbols]);
```

Comment:
I haven't used it before but I think `useMemo` was designed for this.

```typescript
const symbols = useMemo(
  () => symbolDataManager.getSymbols(selectedSymbolGroup),
  [selectedSymbolGroup]
); 
const filteredSymbols = useMemo(
  () => symbolDataManager.filterSymbols(filterText, symbols),
  [filterText, symbols]
);
```
It's nice because it makes it clear that the values of `symbols` and `filteredSymbols` are purely derived from their inputs, rather than having their own independent statefulness.

Reply:
Great improvement, thank you.